core/src/worker/workflow/machines/workflow_machines.rs

```diff
@@ -456,6 +456,31 @@ impl WorkflowMachines {
             .add_lang_used(flags);
     }
 
+    /// Undo a speculative workflow task by resetting to a certain WFT Started ID. This can happen
+    /// when an update request is rejected.
+    pub(crate) fn reset_last_started_id(&mut self, id: i64) {
+        debug!("Resetting back to event id {} due to speculative WFT", id);
+        self.current_started_event_id = id;
+        // This is pretty nasty to just + 1 like this, but, we know WFT complete always follows
+        // WFT started, which the id given to us to reset to must be, and we need to avoid
+        // re-applying the WFT Completed event, so we make sure to consider that processed
+        self.last_processed_event = id + 1;
+        // Then, we have to drop any state machines (which should only be one workflow task machine)
+        // we may have created when servicing the speculative task.
+        // Remove when https://github.com/rust-lang/rust/issues/59618 is stable
+        let remove_these: Vec<_> = self
+            .machines_by_event_id
+            .iter()
+            .filter(|(mid, _)| **mid >= id)
```

Comment:
Nit: Don't think you need the equals because you `id` is a WFT started event so no state machine can be associated with that event